### PR TITLE
feat(algolia-search): allow translating search modal

### PR DESCRIPTION
**Pull request by forresst**
## Pre-flight checklist

- [X] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).

## Motivation

The Algolia Search Modal is not translatable. This should solve the problem.

## Test Plan

~~I'm still trying to figure out how to test the site locally in French... help me~~

Tested locally:

image

### Test links

Deploy preview: https://deploy-preview-_____--docusaurus-2.netlify.app/

## Related issues/PRs

[#4542](https://github.com/facebook/docusaurus/issues/4542)

### CODE REVIEW

File: packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx [hunk 1]
```diff
@@ -221,6 +222,149 @@ function DocSearch({
     description: 'The ARIA label and placeholder for search button',
   });
 
+  const translatedModal: DocSearchTranslations = {
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx:225]: Can we extract all the translations to a separate file, or at least outside the component?

Comment (slorber) [reply packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx:225]: agree 👍 we'll want to reduce the size of this component, and probably later also use some of those translations on the search page (quite old/legacy, but we'll want to refactor/upgrade someday)

File: packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx [hunk 2]
```diff
@@ -221,6 +222,149 @@ function DocSearch({
...
+        cancelButtonText: translate({
+          id: 'theme.SearchModal.searchBox.cancelButtonText',
+          message: 'Cancel',
+          description: 'The label for search box cancel button',
+        }),
+        cancelButtonAriaLabel: translate({
+          id: 'theme.SearchModal.searchBox.cancelButtonAriaLabel',
+          message: 'Cancel',
+          description: 'The ARIA label for search box cancel button',
+        }),
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx:247]: Looks repetitive to me. What are the odds that the user wants different translations for these labels?

Comment (forresst) [reply packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx:247]: OK. I'll change it to one

File: packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx (hunk 1 above)
Comment (Josh-Cena) [reply packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx:225]: Um, it just means copying everything to another file like `src/theme/searchTranslations.ts`, and then importing it as `import translations from "@theme/searchTranslations"`...

**Comment by forresst** [comment]
> > I'm still trying to figure out how to test the site locally in French... help me
> 
> Try `yarn start -l fr`? Default theme translations shouldn't even need to run `write-translations`.

Thank you. But my test doesn't work, I need to look into the details of why it keeps the default values of DocSearch. It's the Modal part (only) that doesn't translate ..... 👀

**Comment by Josh-Cena** [comment]
Did you run `yarn build:packages` or `yarn workspace @docusaurus/theme-algolia-search build` first?

File: packages/docusaurus-theme-search-algolia/src/theme/SearchBar/translations.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  },
+};
+
+const placeholderDocSearchTranslation = translate({
+  id: 'theme.SearchModal.placeholder',
+  message: 'Search docs',
+  description: 'The placeholder of the input of the DocSearch pop-up modal',
+});
+
+export default {docSearchTranslations, placeholderDocSearchTranslation};
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-search-algolia/src/theme/SearchBar/translations.ts:173]:
```suggestion
export {docSearchTranslations, placeholderDocSearchTranslation};
```

**Comment by Josh-Cena** [comment]
Type-checking errors because `@algolia/autocomplete-preset-algolia` is not installed. This only happens now because we never pulled in docsearch's types before. This looks more like a bug on Algolia's side?

**Comment by Josh-Cena** [comment]
cc @shortcuts `@algolia/autocomplete-core` references `@algolia/autocomplete-preset-algolia` in its type definition but doesn't have it as dependency/peer dependency. Maybe `@docsearch/react` need to install it, at least.

**Comment by shortcuts** [comment]
Hey, should be fixed in 3.1.1 for the DocSearch part, I've pinged the Autocomplete team for the `client-search` warning but you shouldn't get an error at least

**Comment by forresst** [comment]
@Josh-Cena @shortcuts It looks good now

**Review by Josh-Cena — APPROVED** [review]
Code LGTM

Not sure about the naming for the new theme module, since these are just the modal translations, not translations for the entire plugin, but not a blocker

**Comment by slorber** [comment]
LGTM 👍  thanks

**Pull request merged**